storage_controller/src/service/safekeeper_service.rs

```diff
@@ -350,6 +353,85 @@ impl Service {
         Ok(())
     }
 
+    /// Perform tenant deletion on safekeepers.
+    pub(super) async fn tenant_delete_safekeepers(
+        self: &Arc<Self>,
+        tenant_id: TenantId,
+    ) -> Result<(), ApiError> {
+        let timeline_list = self
+            .persistence
+            .list_timelines_for_tenant(tenant_id)
+            .await?;
+
+        if timeline_list.is_empty() {
+            // Early exit: the tenant is either empty or not migrated to the storcon yet
+            return Ok(());
+        }
+
+        let timeline_list = timeline_list
+            .into_iter()
+            .map(|timeline| {
+                let timeline_id = TimelineId::from_str(&timeline.timeline_id)
+                    .context("timeline id loaded from db")
+                    .map_err(ApiError::InternalServerError)?;
+                Ok((timeline_id, timeline))
+            })
+            .collect::<Result<Vec<_>, ApiError>>()?;
+
+        // Remove pending ops from db.
+        // We cancel them in a later iteration once we hold the state lock.
+        for (timeline_id, _timeline) in timeline_list.iter() {
+            self.persistence
+                .remove_pending_ops_for_timeline(tenant_id, Some(*timeline_id))
+                .await?;
+        }
+
+        let mut locked = self.inner.write().unwrap();
+
+        // The list of safekeepers that have any of the timelines
+        let mut sk_list = HashSet::new();
+
+        // List all pending ops for all timelines, cancel them
+        for (timeline_id, timeline) in timeline_list.iter() {
+            let sk_iter = timeline
+                .sk_set
+                .iter()
+                .chain(timeline.new_sk_set.iter().flatten())
+                .map(|id| NodeId(*id as u64));
+            for sk_id in sk_iter.clone() {
+                locked
+                    .safekeeper_reconcilers
+                    .cancel_reconciles_for_timeline(sk_id, tenant_id, Some(*timeline_id));
+            }
+            sk_list.extend(sk_iter);
+        }
+
+        // unwrap is safe: we return above for an empty timeline list
+        let max_generation = timeline_list
```

Comment:
For deletion generation doesn't really make sense; let's use null/none instead.

Reply:
I think having a dummy generation in the table is better than making something nullable in order to match what's stored 100% with what is being needed for the op. you can see in this PR how much additional code a nullable timeline id caused :).

Reply:
Well, rust is more annoying that python in this sense, yes. But this is a hack, untrivially calculating something unclear instead of basic null. Moreover, again there is a very unlikely race that this max gets stale if new operation is concurrently queued; in this case persistence of deletion op might get ignored. Better avoid hacks.

Reply:
alright, let's use null, but can we make this as a followup? I don't want to add a diesel migration just for the generation, and I don't want to add any foreign key constraints to this PR, to keep it only deletion related.

Reply:
no objections